strax/processing/hitlets.py

```diff
@@ -133,219 +136,98 @@ def refresh_hit_to_hitlets(hits, hitlets):
-
-            if counter > TRIAL_COUNTER_NEIGHBORING_RECORDS:
-                print(ind, last_ri)
-                raise RuntimeError('Tried too often to find correct record_i.')
+    # Numba will not raise any exceptions if to_pe is too short, leading
+    # to strange bugs. This check is somewhat not complete, but should
+    # be sufficient in most cases.
+    to_pe_has_wrong_shape = len(to_pe) < hitlets['channel'].max()
+    if to_pe_has_wrong_shape:
+        raise ValueError('"to_pe" has a wrong shape. Array index must'
+                         ' match channel numbers.')
+
+    hitelts_is_single_row = isinstance(hitlets, np.void)
+    if hitelts_is_single_row:
+        # A structured array becomes void type if a single row is called,
+        # e.g. hitlets[0] which does not work in numba while, hitlets[:1]
+        # does. So we have to convert the row into the correct format first.
+        hitlets = np.array([hitlets])
+
+    data_field_in_hitlets = 'data' in hitlets.dtype.names
+    if data_field_in_hitlets:
+        data_is_not_empty = np.any(hitlets['data'] != 0)
+        if data_is_not_empty:
+            raise ValueError('The data field of hitlets must be empty!')
+
+        data_field_not_long_enough = len(hitlets[0]['data']) < hitlets['length'].max()
+        if data_field_not_long_enough:
+            raise ValueError('The data field must be as large as the longest hitlet in our data.')
+
+        hitlets_with_data_field = hitlets
+    else:
+        n_samples = max(100, hitlets['length'].max())
+        hitlets_with_data_field = np.zeros(len(hitlets), strax.hitlet_with_data_dtype(n_samples))
+        strax.copy_to_buffer(hitlets,
+                             hitlets_with_data_field,
+                             '_copy_hitlets_to_hitlets_width_data')
+
+    _get_hitlets_data(hitlets_with_data_field, records, to_pe)
+    return hitlets_with_data_field
+
+
+@numba.jit(nopython=True, nogil=True, cache=True)
+def _get_hitlets_data(hitlets, records, to_pe):
+    rranges = _touching_windows(records['time'],
+                                strax.endtime(records),
+                                hitlets['time'],
+                                strax.endtime(hitlets))
+
+    for i, h in enumerate(hitlets):
+        recorded_samples_offset = 0
+        n_recorded_samples = 0
+        is_first_record = True
+        for ind, r_ind in enumerate(range(rranges[i][0], rranges[i][1])):
+            r = records[r_ind]
+            if r['channel'] != h['channel']:
+                continue
+
+            (r_start, r_end), (h_start, h_end) = strax.overlap_indices(
+                r['time'] // r['dt'],
+                r['length'],
+                h['time'] // h['dt'],
+                h['length'])
+
+            if is_first_record:
+                is_first_record = False
+                recorded_samples_offset = h_start
+            h_start -= recorded_samples_offset
+            h_end -= recorded_samples_offset
```

Comment:
This looks a bit funny, you do this for *each* record, not only the first?

Reply:
okay, yes and `h_start` gets reset every record, whereas `recorded_samples_offset`, got ya